flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/common/WindowStartEndPropertiesRule.scala

```diff
@@ -54,34 +56,75 @@ class WindowStartEndPropertiesRule
     val project = call.rel(0).asInstanceOf[LogicalProject]
     val innerProject = call.rel(1).asInstanceOf[LogicalProject]
     val agg = call.rel(2).asInstanceOf[LogicalWindowAggregate]
+    val window = agg.getWindow
 
-    // Retrieve window start and end properties
+    val isRowtime = isRowtimeAttribute(window.timeAttribute)
+    val isProctime = isProctimeAttribute(window.timeAttribute)
+
+    val startEndProperties = Seq(
+      NamedWindowProperty("w$start", WindowStart(window.aliasAttribute)),
+      NamedWindowProperty("w$end", WindowEnd(window.aliasAttribute)))
+
+    // allow rowtime/proctime for rowtime windows and proctime for proctime windows
+    val timeProperties = if (isRowtime) {
+      Seq(
+        NamedWindowProperty("w$rowtime", RowtimeAttribute(window.aliasAttribute)),
```

Comment:
We have so many `w$...` here, can we create the String constants and reference them ?